pco/src/float_decomp_utils.rs

```diff
@@ -0,0 +1,58 @@
+use std::cmp;
+
+use crate::constants::Bitlen;
+use crate::data_types::FloatLike;
+use crate::sampling;
+
+#[inline(never)]
+pub(crate) fn join_latents<F: FloatLike>(k: Bitlen, primary: &mut [F::L], secondary: &[F::L]) {
+  for (y_and_dst, &m) in primary.iter_mut().zip(secondary.iter()) {
+    assert!(m >> k == 0, "Invalid input to FloatDecomp: m must be a k-bit integer")
+    *y_and_dst = F::from_bits((*y_and_dst << k) + m).to_latent_ordered()
```

Comment:
So I thought about it and believe we want to do this a bit differently. There's quite a lot we want out of these formulas:

* The number should be monotonically increasing with the primary latent. This ensures that delta encoding the primary stream of latents makes sense and increases the smoothness of the latent distribution.
* It should be possible to write the formula in a performant way (already great!).
* Ideally the formula is easily explainable, since this will be a lasting part of the format.

Mainly the issue with the current formula is that it has a big jump and changes monotonicity direction around 0.0. I'm thinking we should do something more like this:
```rust
let quantized = F::from_latent_ordered(*y_and_dst << k);
F::from_latent_bits(quantized.to_latent_bits() + m)
```

This is order-preserving in the first latent, still pretty performant (not as much as your formula, but I think that tradeoff is well worth it for the compression ratio), and only a little bit hard to explain: "The first latent is the number of quantums away from -inf, and the second is the number of ULPs to increase magnitude by".

Do you agree?

Reply:
Ah, I didn't think of that.  Yeah, it makes sense that we want the `split_latents` mapping, which in general already assumes that the second component is compressible, to try its best to make the first component compressible as well.

I agree there is value to having the formula be explainable in terms of abstract numbers (as opposed to being entirely based on peculiarities of the floating point format).  At this point I'm not able to assess how that compares to the value of keeping the first component of `split_latents` compressible, but I definitely see both.

I'm not sure I understand the comment about monotonicity and smoothness.  I would think that monotonicity alone doesn't imply that the pushforward of a smooth distribution by `split_latents` is also smooth, right?

Reply:
Oh this is great, thank you.  I'll think about it in more detail this weekend and I bet I'll learn from it.

Reply:
Ok, I think I understand your previous comment, except a couple of details (mainly, intuitions about how benchmarks will perform) which I don't need to fully understand now.  I agree with your suggestion and will make the change, thanks!